data/tokenizer.py

```diff
@@ -120,3 +128,75 @@ def untokenize(self, tokens: list[str]) -> pd.DataFrame:
         df = pd.DataFrame(velocities, columns=["velocity"])
 
         return df
+
+
+class DstartEncoder(MidiEncoder):
+    def __init__(self, bins: int = 200):
+        super().__init__()
+        self.specials = ["<s>", "</s>", "<blank>"]
+        self.bins = bins
+        # Make a copy of special tokens ...
+        self.vocab = list(self.specials)
+        # ... and add velocity tokens
+        self._build_vocab()
+        self._bin_edges = self._load_bin_edges()
+        self.bin_to_dstart = []
+        self._build_dstart_decoder()
+        self.token_to_id = {token: it for it, token in enumerate(self.vocab)}
+
+    def _load_bin_edges(self):
+        artifacts_path = to_absolute_path("artifacts/bin_edges.yaml")
+        with open(artifacts_path, "r") as f:
+            bin_edges = yaml.safe_load(f)
+
+        dstart_bin_edges = bin_edges["dstart"][self.bins]
+        return dstart_bin_edges
+
+    def _build_vocab(self):
+        self.vocab += [str(possible_bin) for possible_bin in range(self.bins)]
+
+    def _build_dstart_decoder(self):
+        self.bin_to_dstart = []
+        for it in range(1, len(self._bin_edges)):
+            dstart = (self._bin_edges[it - 1] + self._bin_edges[it]) / 2
+            self.bin_to_dstart.append(dstart)
+
+        last_dstart = 2 * self._bin_edges[-1]
+        self.bin_to_dstart.append(last_dstart)
+
+    @property
+    def vocab_size(self) -> int:
+        return len(self.vocab)
+
+    def unquantized_start(self, dstart_bins: pd.Series) -> pd.Series:
+        quant_dstart = [self.bin_to_dstart[it] for it in dstart_bins]
+        start = pd.Series(quant_dstart).cumsum().shift(1).fillna(0)
+
+        return start
+
+    def quantize(self, start: list[float]) -> list[int]:
+        dstart = []
+        for it in range(len(start) - 1):
+            dstart.append(start[it + 1] - start[it])
+        dstart.append(0)
+
+        dstart_bins = np.digitize(dstart, self._bin_edges) - 1
+
+        return dstart_bins
+
+    def tokenize(self, record: dict) -> list[str]:
+        # TODO I don't love the idea of adding tokens durint *tokenize* call
+        # If we want to pretend that our midi sequences have start and finish
+        # we should take care of that before we get here :alarm:
+        dstart_bins = self.quantize(record["start"])
+
+        # get tokens from quantized data
+        tokens = [str(dstart_bin) for dstart_bin in dstart_bins]
+        tokens = ["<s>"] + tokens + ["</s>"]
+        return tokens
+
+    def untokenize(self, tokens: list[str]) -> pd.DataFrame:
+        dstarts = [int(token) for token in tokens if token not in self.specials]
+        df = pd.DataFrame(dstarts, columns=["dstart_bin"])
+
+        return df
```

Comment:
```suggestion
    def untokenize(self, tokens: list[str]) -> list[float]:
        dstarts = [int(token) for token in tokens if token not in self.specials]

        return dstarts
```

I'll revisit this after running the code more, but I'm guessing returning a list of floats here will make inference experiments easier to follow

Reply:
How about we try and change it in VelocityEncoder as well then? Just so child objects implement methods the same way

Reply:
Definitely!

Reply:
QuantizedMidiEncoder untokenize would still need to return a dataframe though, as it untokenizes into four keys.

Reply:
I think that's ok - it encodes and returns a complex structure so it seems justified

Reply:
I found a problem - righ now we take greedy decode output - which is baisically output token ids - and use untokenize function as if they were tokens (strings)
We should convert output to tokens first or create a function to decode token ids directly.

Reply:
Sorry I don't follow - how can we use untokenize on token ids as if they were tokens?

Reply:
Right now we do smtm like this:
Model output --untokenize—> sequence (all in generate output function)

We should do is
Model output -> vocab[id] for id in model output -> untokenize -> sequence

I do not know why when passing tensor of ints into untokenize we do not get error with type hints, as the argument for untokenize should be a list of strings

Reply:
Can you link to a place in the code where this is happening? I just took a random look at one of the velocity dashboards `dashboard/velocity/predict_piece.py:77`, and there is vocab after model:

```py
out_tokens = [tgt_encoder.vocab[x] for x in predicted_token_ids if x != pad_idx]
```

Reply:
yup, it works great here